lib/chef/resource/lwrp_base.rb

```diff
@@ -35,32 +35,64 @@ class LWRPBase < Resource
 
       # Evaluates the LWRP resource file and instantiates a new Resource class.
       def self.build_from_file(cookbook_name, filename, run_context)
-        resource_class = nil
-        rname = filename_to_qualified_string(cookbook_name, filename)
-
-        class_name = convert_to_class_name(rname)
-        if Resource.const_defined?(class_name, false)
-          Chef::Log.info("#{class_name} light-weight resource is already initialized -- Skipping loading #{filename}!")
-          Chef::Log.debug("Overriding already defined LWRPs is not supported anymore starting with Chef 12.")
-          resource_class = Resource.const_get(class_name)
-        else
-          resource_class = Class.new(self)
+        resource_name = filename_to_qualified_string(cookbook_name, filename)
+
+        node = run_context ? run_context.node : Chef::Node.new
+        existing_class = Chef::ResourceResolver.new(node, resource_name).resolve
+
+        if existing_class
+          Chef::Log.info("Skipping LWRP resource #{filename} from cookbook #{cookbook_name}.  #{existing_class} already defined.")
+          Chef::Log.debug("Overriding resources with LWRPs is not supported anymore starting with Chef 12.")
+          return existing_class
+        end
 
-          Chef::Resource.const_set(class_name, resource_class)
-          resource_class.resource_name = rname
-          resource_class.run_context = run_context
-          resource_class.class_from_file(filename)
+        # We load the class first to give it a chance to set its own name
+        resource_class = Class.new(self)
+        resource_class.resource_name = resource_name
+        resource_class.run_context = run_context
+        resource_class.class_from_file(filename)
 
-          Chef::Log.debug("Loaded contents of #{filename} into a resource named #{rname} defined in Chef::Resource::#{class_name}")
+        # Respect resource_name set inside the LWRP
+        resource_class.instance_eval do
+          define_method(:to_s) do
+            "LWRP #{resource_name} from cookbook #{cookbook_name}"
+          end
+          define_method(:inspect) { to_s }
         end
 
+        Chef::Log.debug("Loaded contents of #{filename} into #{resource_class}")
+
+        create_deprecated_class_in_chef_resource(resource_class)
+
         resource_class
       end
 
-      # Set the resource name for this LWRP
+      def self.create_deprecated_class_in_chef_resource(resource_class)
+        # Create a class in Chef::Resource::MyResource with deprecation
+        # warnings if you try to instantiate or inherit it (for Chef 12 compatibility)
+        class_name = convert_to_class_name(resource_class.resource_name)
+        if Chef::Resource.const_defined?(class_name, false)
+          Chef::Log.debug "#{class_name} already exists!  Cannot create deprecation class #{resource_class}"
+        else
+          Chef::Resource.const_set(class_name, Class.new(resource_class) do
```

Comment:
What did you have in mind?  I'm not sure we'd want to make it *always* create a new Chef::Resource class in `const_missing`, and I'm not sure what else I could do in there.